commons/zenoh-keyexpr/src/key_expr/borrowed.rs

```diff
@@ -286,11 +290,247 @@ impl keyexpr {
     pub unsafe fn from_slice_unchecked(s: &[u8]) -> &Self {
         core::mem::transmute(s)
     }
-    pub fn chunks(&self) -> impl Iterator<Item = &Self> + DoubleEndedIterator {
-        self.split('/').map(|c| unsafe {
-            // Any chunk of a valid KE is itself a valid KE => we can safely call the unchecked constructor.
-            Self::from_str_unchecked(c)
-        })
+    pub const fn chunks(&self) -> Chunks {
+        Chunks {
+            inner: self.as_str(),
+        }
+    }
+    pub(crate) fn next_delimiter(&self, i: usize) -> Option<usize> {
+        self.as_str()
+            .get(i + 1..)
+            .and_then(|s| s.find('/').map(|j| i + 1 + j))
+    }
+    pub(crate) fn previous_delimiter(&self, i: usize) -> Option<usize> {
+        self.as_str().get(..i).and_then(|s| s.rfind('/'))
+    }
+    pub(crate) fn first_byte(&self) -> u8 {
+        unsafe { *self.as_bytes().get_unchecked(0) }
+    }
+    pub(crate) fn iter_splits_ltr(&self) -> SplitsLeftToRight {
+        SplitsLeftToRight {
+            inner: self,
+            index: 0,
+        }
+    }
+    pub(crate) fn iter_splits_rtl(&self) -> SplitsRightToLeft {
+        SplitsRightToLeft {
+            inner: self,
+            index: self.len(),
+        }
+    }
+}
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub(crate) struct SplitsLeftToRight<'a> {
+    inner: &'a keyexpr,
+    index: usize,
+}
+impl<'a> SplitsLeftToRight<'a> {
+    fn right(&self) -> &'a str {
+        &self.inner[self.index + ((self.index != 0) as usize)..]
+    }
+    fn left(&self, followed_by_double: bool) -> &'a str {
+        &self.inner[..(self.index + ((self.index != 0) as usize + 2) * followed_by_double as usize)]
+    }
+}
+impl<'a> Iterator for SplitsLeftToRight<'a> {
+    type Item = (&'a keyexpr, &'a keyexpr);
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.index < self.inner.len() {
+            false => None,
+            true => {
+                let right = self.right();
+                let double_wild = right.starts_with("**");
+                let left = self.left(double_wild);
+                self.index = if left.is_empty() {
+                    self.inner.next_delimiter(0).unwrap_or(self.inner.len())
+                } else {
+                    self.inner
+                        .next_delimiter(left.len())
+                        .unwrap_or(self.inner.len() + (left.len() == self.inner.len()) as usize)
+                };
+                if left.is_empty() {
+                    self.next()
+                } else {
+                    (!right.is_empty()).then(|| unsafe {
```

Comment:
It's good practice to comment on why calling `unsafe` is actual *safe* in the current implementation.
That would ease review and debugging. This same very comment applies to all other `unsafe` blocks.